File: Books/A_Comprehensive_Approach_to_ML_and_AI/matlab/gbt_tutorial.m
close all; clear all; clc; 

% Example usage:
rng(1); % For reproducibility
% Generate synthetic data for regression
X = randn(100,2);
y = sin(X(:,1)) + 0.5*X(:,2) + randn(100,1)*0.1;  % Some nonlinear function

M = 50;      % Number of boosting iterations
eta = 0.1;   % Learning rate
F = gradient_boosting_trees(X, y, M, eta);

% Compute predictions for training data
y_pred = F(X);

% Plot the true vs predicted values
figure;
scatter(y, y_pred, 'bo');
xlabel('True Values');
ylabel('Predicted Values');
title('Gradient Boosting Trees Regression');
grid on;

% save_all_figs_OPTION('results/gbt','png',1)


function F = gradient_boosting_trees(X, y, M, eta)
    % GRADIENTBOOSTINGTREES performs gradient boosting for regression.
    %   X: m x n data matrix (each row is a data point)
    %   y: m x 1 target vector
    %   M: number of boosting iterations
    %   eta: learning rate
    %   F: final predictive model (function handle)
    
    m = size(X,1);
    
    % Initialize model: for squared error loss, the best constant is the mean.
    F_pred = mean(y) * ones(m, 1);
    models = cell(M,1);
    gammas = zeros(M,1);
    
    for iter = 1:M
        % Compute residuals (for squared error loss, residuals are y - F_pred)
        residuals = y - F_pred;
        
        % Fit a simple regression tree to residuals (from scratch)
        tree = buildTree(X, residuals, 5, 3); % 5: maxDepth, 3: minLeaf (example parameters)
        models{iter} = tree;
        
        % Compute optimal gamma (for squared error, optimal gamma is 1)
        gamma = 1;  % For squared error loss, this is optimal.
        gammas(iter) = gamma;
        
        % Update prediction
        update = predictTree(tree, X);
        F_pred = F_pred + eta * gamma * update;
    end
    
    % Return final model as a function handle that aggregates predictions
    F = @(X_new) aggregatePrediction(X_new, models, gammas, eta, mean(y));
end

function tree = buildTree(X, y, maxDepth, minLeaf)
    % BUILDTREE builds a regression tree recursively.
    if (maxDepth == 0) || (length(y) < minLeaf) || (std(y)==0)
        tree.isLeaf = true;
        tree.prediction = mean(y);
        return;
    end
    
    [bestFeature, bestThreshold, bestGain] = chooseBestSplit(X, y);
    if bestGain <= 0
        tree.isLeaf = true;
        tree.prediction = mean(y);
        return;
    end
    
    tree.isLeaf = false;
    tree.feature = bestFeature;
    tree.threshold = bestThreshold;
    
    leftIdx = X(:, bestFeature) <= bestThreshold;
    rightIdx = ~leftIdx;
    
    tree.left = buildTree(X(leftIdx, :), y(leftIdx), maxDepth-1, minLeaf);
    tree.right = buildTree(X(rightIdx, :), y(rightIdx), maxDepth-1, minLeaf);
end

function [bestFeature, bestThreshold, bestGain] = chooseBestSplit(X, y)
    % CHOOSEBESTSPLIT finds the best feature and threshold using squared error reduction.
    [numSamples, numFeatures] = size(X);
    bestGain = -inf;
    bestFeature = 0;
    bestThreshold = 0;
    currentError = sum((y - mean(y)).^2);
    
    for feature = 1:numFeatures
        thresholds = unique(X(:, feature));
        for t = thresholds'
            leftIdx = X(:, feature) <= t;
            rightIdx = ~leftIdx;
            if isempty(y(leftIdx)) || isempty(y(rightIdx))
                continue;
            end
            errorLeft = sum((y(leftIdx) - mean(y(leftIdx))).^2);
            errorRight = sum((y(rightIdx) - mean(y(rightIdx))).^2);
            gain = currentError - (length(y(leftIdx))/numSamples * errorLeft + ...
                   length(y(rightIdx))/numSamples * errorRight);
            if gain > bestGain
                bestGain = gain;
                bestFeature = feature;
                bestThreshold = t;
            end
        end
    end
end

function y_pred = predictTree(tree, X)
    % PREDICTTREE predicts responses for X using the regression tree.
    m = size(X,1);
    y_pred = zeros(m,1);
    for i = 1:m
        y_pred(i) = traverseTree(tree, X(i,:));
    end
end

function prediction = traverseTree(tree, x)
    if tree.isLeaf
        prediction = tree.prediction;
    else
        if x(tree.feature) <= tree.threshold
            prediction = traverseTree(tree.left, x);
        else
            prediction = traverseTree(tree.right, x);
        end
    end
end

function pred = aggregatePrediction(X_new, models, gammas, eta, initPrediction)
    % AGGREGATEPREDICTION aggregates the predictions of all trees.
    m = size(X_new,1);
    pred = initPrediction * ones(m,1);
    for i = 1:length(models)
        update = predictTree(models{i}, X_new);
        pred = pred + eta * gammas(i) * update;
    end
end

File: Books/A_Comprehensive_Approach_to_ML_and_AI/R/gbt_tutorial.R
# gbt_tutorial.R
# Gradient Boosting Trees Tutorial
#
# This script demonstrates a simplified gradient boosting trees algorithm
# for binary classification using logistic loss.
#
# We:
#   1. Generate synthetic 2D data:
#         - Class 1: centered at (2,2)
#         - Class 0: centered at (-2,-2)
#   2. Train an ensemble of regression stumps sequentially by fitting the negative gradient
#      (residuals) of the logistic loss.
#   3. Compute predictions via an additive model and convert to probabilities using the sigmoid.
#   4. Visualize the decision boundary and report training accuracy.

rm(list = ls())
graphics.off()
set.seed(1)

## 1. Generate Synthetic Data
N <- 100
# Class 1: centered at (2,2)
X1 <- matrix(rnorm(N * 2, mean = 2), ncol = 2)
# Class 0: centered at (-2,-2)
X0 <- matrix(rnorm(N * 2, mean = -2), ncol = 2)
X <- rbind(X1, X0)
y <- c(rep(1, N), rep(0, N))

## Sigmoid Function
sigmoid <- function(x) { 1 / (1 + exp(-x)) }

## 2. Train Gradient Boosting Trees Model
gradientBoostingTrain <- function(X, y, T, eta) {
  n <- nrow(X)
  F <- rep(0, n)  # Initial model output F(x)=0
  models <- vector("list", T)
  gammas <- numeric(T)
  
  for (t in 1:T) {
    p <- sigmoid(F)
    residuals <- y - p
    stump <- stumpRegTrain(X, residuals)
    h <- stumpRegPredict(stump, X)
    gamma <- 1  # optimal for squared error loss
    gammas[t] <- gamma
    F <- F + eta * gamma * h
    models[[t]] <- list(stump = stump, coef = eta * gamma)
  }
  list(models = models)
}

stumpRegTrain <- function(X, r) {
  # Train a regression stump that minimizes squared error.
  n <- nrow(X); d <- ncol(X)
  bestLoss <- Inf
  bestFeature <- 1; bestThreshold <- 0; bestC1 <- 0; bestC2 <- 0
  for (j in 1:d) {
    thresholds <- unique(X[, j])
    for (thresh in thresholds) {
      left <- X[, j] < thresh
      right <- !left
      if (sum(left) == 0 || sum(right) == 0) next
      c1 <- mean(r[left])
      c2 <- mean(r[right])
      loss <- sum((r[left] - c1)^2) + sum((r[right] - c2)^2)
      if (loss < bestLoss) {
        bestLoss <- loss
        bestFeature <- j
        bestThreshold <- thresh
        bestC1 <- c1
        bestC2 <- c2
      }
    }
  }
  list(feature = bestFeature, threshold = bestThreshold, c1 = bestC1, c2 = bestC2)
}

stumpRegPredict <- function(stump, X) {
  # Predict with the regression stump.
  preds <- ifelse(X[, stump$feature] < stump$threshold, stump$c1, stump$c2)
  preds
}

## 3. Prediction Function for Gradient Boosting Model
gradientBoostingPredict <- function(X, modelStruct) {
  n <- nrow(X)
  F <- rep(0, n)
  for (m in modelStruct$models) {
    F <- F + m$coef * stumpRegPredict(m$stump, X)
  }
  sigmoid(F)
}

gbt_model <- gradientBoostingTrain(X, y, T = 50, eta = 0.1)
y_pred <- gradientBoostingPredict(X, gbt_model)
y_pred_class <- ifelse(y_pred >= 0.5, 1, 0)
train_acc <- mean(y_pred_class == y) * 100
cat(sprintf("Gradient Boosting Trees Training Accuracy: %.2f%%\n", train_acc))

## 4. Visualize Decision Boundary
x_min <- min(X[,1]) - 1; x_max <- max(X[,1]) + 1
y_min <- min(X[,2]) - 1; y_max <- max(X[,2]) + 1
xx <- seq(x_min, x_max, length.out = 200)
yy <- seq(y_min, y_max, length.out = 200)
gridPoints <- as.matrix(expand.grid(xx, yy))
preds_grid <- gradientBoostingPredict(gridPoints, gbt_model)
preds_grid_mat <- matrix(preds_grid, nrow = 200, ncol = 200)

par(mfrow = c(2,1))
plot(X[y==1,1], X[y==1,2], col = "blue", pch = 16, xlab = "Feature 1", ylab = "Feature 2",
     main = "Gradient Boosting Trees Classification")
points(X[y==0,1], X[y==0,2], col = "red", pch = 16)
contour(xx, yy, preds_grid_mat, levels = 0.5, add = TRUE, col = "black", lwd = 2)
